compiler/rustc_hir_typeck/src/mem_categorization.rs

```diff
@@ -715,13 +715,19 @@ impl<'a, 'tcx> MemCategorizationContext<'a, 'tcx> {
                 self.cat_pattern_(place_with_id, subpat, op)?;
             }
 
-            PatKind::Box(subpat) | PatKind::Ref(subpat, _) | PatKind::Deref(subpat) => {
+            PatKind::Box(subpat) | PatKind::Ref(subpat, _) => {
                 // box p1, &p1, &mut p1. we can ignore the mutability of
                 // PatKind::Ref since that information is already contained
                 // in the type.
                 let subplace = self.cat_deref(pat, place_with_id)?;
                 self.cat_pattern_(subplace, subpat, op)?;
             }
+            PatKind::Deref(subpat) => {
+                let ty = self.pat_ty_adjusted(subpat)?;
+                // A deref pattern generates a temporary.
+                let place = self.cat_rvalue(pat.hir_id, ty);
```

Comment:
The temporary has the type of a reference to ty as the place, but also this is only for closures so temporaries are going to be ignored. There might need to be some more significant changes here for closure captures to work correctly with deref patterns.

Reply:
Oops well spotted.

What would "work correctly" look like? Presumably we can only capture a ref to the smart pointer, right? That works today:
```rust
fn main() {
    let b = Box::new("aaa".to_string());
    let f = || {
        let deref!(ref s) = b else { unreachable!() };
        assert_eq!(s.len(), 3);
    };
    assert_eq!(b.len(), 3);
    f();
}
```

Reply:
Ah, `deref!(x) = &mut b` works, but `deref!(ref mut x) = b` only captures `&b` for some reason

Reply:
I think that one or both of `walk_pat` and `maybe_read_scrutinee` in `expr_use_visitor.rs` need to handle `PatKind::Deref`.

Reply:
Thank you! That worked